File: Graphs/820-find-eventual-safe-states/find-eventual-safe-states.py
class Solution:
    def eventualSafeNodes(self, graph):
        n = len(graph)
        ultagraph = [[] for _ in range(n)]
        indegree = [0] * n
        ans = []

        for i in range(n):
            for neighbor in graph[i]:
                ultagraph[neighbor].append(i)
                indegree[i] += 1

        q = deque()
        for i in range(n):
            if indegree[i] == 0:
                q.append(i)

        while q:
            node = q.popleft()
            ans.append(node)

            for neighbor in ultagraph[node]:
                indegree[neighbor] -= 1
                if indegree[neighbor] == 0:
                    q.append(neighbor)

        ans.sort()
        return ans

File: Graphs/820-find-eventual-safe-states/find-eventual-safe-states.cpp
class Solution {
public:
    bool dfs(vector<vector<int>>& graph, vector<int>& vis, vector<int>& ans,
             int node) {
        if (vis[node] == 2)
            return true;

        if(vis[node] == 1)
            return false;

        vis[node] = 1;

        for (auto i : graph[node]) {
            if (!dfs(graph, vis, ans, i))
                return false;
        }

        vis[node] = 2;
        ans.push_back(node);
        return true;
    }

    vector<int> eventualSafeNodes(vector<vector<int>>& graph) {
        int n = graph.size();
        vector<int> vis(n, 0);
        vector<int> ans;

        for (int i = 0; i < n; i++) {
            if (vis[i] == 0)
                dfs(graph, vis, ans, i);
        }

        sort(ans.begin(), ans.end());
        return ans;
    }
};